test/src/utils/lock.test.ts

```diff
@@ -0,0 +1,179 @@
+import "mocha";
+import { expect } from "chai";
+import { Lock, LockManager } from "../../../src/utils/lock";
+import { wait } from "../../../src/utils";
+import { ApplicationError } from "../../../src/dataEntities";
+
+describe("Lock", () => {
+    it("updates 'locked' correctly", async () => {
+        const lock = new Lock();
+
+        await lock.acquire();
+
+        expect(lock.locked, "set locked to true after acquire").to.be.true;
+
+        lock.release();
+
+        expect(lock.locked, "set locked to false after release").to.be.false;
+    });
+
+    it("gives the lock to the waiters in the right order", async () => {
+        const lock = new Lock();
+
+        const callOrder: number[] = [];
+        const p1 = lock.acquire().then(async () => {
+            callOrder.push(1);
+            await wait(30); // make sure to take some time before releasing
+            lock.release();
+        });
+
+        const p2 = lock.acquire().then(() => {
+            callOrder.push(2);
+            lock.release();
+        });
+
+        const p3 = lock.acquire().then(() => {
+            callOrder.push(3);
+            lock.release();
+        });
+
+        await Promise.all([p1, p2, p3]);
+
+        expect(callOrder).to.deep.equal([1, 2, 3]);
+    });
+
+    it("throws ApplicationError if released when not locked", () => {
+        const lock = new Lock();
+        expect(() => lock.release()).to.throw(ApplicationError);
+    });
+
+    it("throws ApplicationError if released two times but acquired only once", () => {
+        const lock = new Lock();
+        lock.acquire();
+        lock.release();
+        expect(() => lock.release()).to.throw(ApplicationError);
+    });
+});
+
+describe("LockManager", () => {
+    it("acquire can acquire a key", () => {
+        const lockManager = new LockManager();
+        return expect(lockManager.acquire("key")).to.be.fulfilled;
+    });
+
+    it("acquire can acquire two different keys", () => {
+        const lockManager = new LockManager();
+        const p1 = lockManager.acquire("key");
+        const p2 = lockManager.acquire("anotherKey");
+        return expect(Promise.all([p1, p2])).to.be.fulfilled;
+    });
+
+    it("acquire cannot acquire a key again before release", async () => {
+        const lockManager = new LockManager();
+        await lockManager.acquire("key");
+        let secondLockAcquired = false;
+        lockManager.acquire("key").then(() => {
+            secondLockAcquired = true;
+        });
+        await wait(30);
+        expect(secondLockAcquired).to.be.false;
+    });
+
+    it("acquire cannot acquire a key again before release", async () => {
```

Comment:
This test has a duplicate name

Reply:
Removed, also the tested functionality was redundant (essentially equal to two tests before).